gm8emulator/src/game/recording/console.rs

```diff
@@ -0,0 +1,175 @@
+use imgui::ListBox;
+
+use crate::{
+    game::recording::window::{ Window, Openable, EmulatorContext },
+    gml::{ Context, Value }
+};
+
+pub struct ConsoleWindow {
+    input_string: String,
+    gml_context: Context,
+    output: Vec<String>,
+
+    scroll_to_bottom: bool,
+    run_code: bool,
+    last_frame: usize,
+    last_rerecords: u64,
+
+    is_open: bool,
+    id: usize,
+}
+
+impl Openable<Self> for ConsoleWindow {
+    fn window_name() -> &'static str {
+        "Console"
+    }
+
+    fn open(id: usize) -> Self {
+        let mut new_console = Self::new();
+        new_console.id = id;
+
+        new_console
+    }
+}
+impl Window for ConsoleWindow {
+    fn stored_kind(&self) -> Option<super::WindowKind> {
+        Some(super::WindowKind::Console(self.id))
+    }
+
+    fn window_id(&self) -> usize {
+        self.id
+    }
+
+    fn name(&self) -> String {
+        format!("Console {}", self.id+1)
+    }
+
+    fn show_window(&mut self, info: &mut EmulatorContext) {
+        let EmulatorContext {
+            config,
+            frame,
+            game,
+            keybindings,
+            clean_state,
+            ..
+        } = info;
+
+        frame.window(self.name())
+            .opened(&mut self.is_open)
+            .position([100.0, 100.0], imgui::Condition::FirstUseEver)
+            .size([600.0, 250.0], imgui::Condition::FirstUseEver)
+            .build(|| {
+                let window_size = frame.window_size();
+                let content_position = frame.window_content_region_min();
+                ListBox::new("GMLConsoleOutput")
+                .size([window_size[0]-content_position[0]*2.0, window_size[1] - 60.0])
+                .build(frame, || {
+                    for text in &self.output {
+                        frame.text(&text);
+                    }
+                    if self.scroll_to_bottom {
+                        self.scroll_to_bottom = false;
+                        frame.set_scroll_here_y_with_ratio(1.0);
+                    }
+                });
+
+                // width = window width - padding - (width of checkbox + item x spacing) - (width of run button + item x spacing)
+                let width = window_size[0] - content_position[0] * 2.0 - 58.0 - 28.0;
```

Comment:
These repeating 58 and 28 here and below look like some cursed magic numbers.
Could we better obtain them from somewhere or at least put them into distinct constants?